# **This notebook splits the dataset into a training and testing data. Training data gets stored in the file Train_Test_Data_Airport_Flight_Delays.pkl and Test data (unseen) data get stored in the file Validation_Data_Airport_Flight_Delays.pkl**

In [ ]:
!pip3 install pickle5

import pickle5 as pickle

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 256 kB 25.3 MB/s 


In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
path = 'drive/MyDrive/categorical_features_data_flightdelay.pkl'

with open(path, "rb") as dt:
  df = pickle.load(dt)


In [ ]:
 y = df['DEP_DEL15']
 X = df.drop('DEP_DEL15',axis = 1)


In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_validate, y_train, y_validate = train_test_split(X, y, test_size=0.30, random_state=2022, stratify = y)

In [ ]:
df_validate = pd.concat([X_validate, y_validate], axis = 1)
df_test = pd.concat([X_train, y_train],axis = 1)
df_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4542343 entries, 2895827 to 5508875
Data columns (total 20 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   MONTH                          int64  
 1   DAY_OF_WEEK                    int64  
 2   DISTANCE_GROUP                 int64  
 3   SEGMENT_NUMBER                 int64  
 4   CONCURRENT_FLIGHTS             int64  
 5   CARRIER_NAME_INT               int32  
 6   AIRLINE_AIRPORT_FLIGHTS_MONTH  int64  
 7   AVG_MONTHLY_PASS_AIRPORT       int64  
 8   AVG_MONTHLY_PASS_AIRLINE       int64  
 9   FLT_ATTENDANTS_PER_PASS        float64
 10  GROUND_SERV_PER_PASS           float64
 11  PLANE_AGE                      int64  
 12  LATITUDE                       float64
 13  LONGITUDE                      float64
 14  PRCP                           float64
 15  SNOW                           float64
 16  SNWD                           float64
 17  TMAX                           float64
 

In [ ]:
df_validate.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1946719 entries, 2192175 to 6328489
Data columns (total 20 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   MONTH                          int64  
 1   DAY_OF_WEEK                    int64  
 2   DISTANCE_GROUP                 int64  
 3   SEGMENT_NUMBER                 int64  
 4   CONCURRENT_FLIGHTS             int64  
 5   CARRIER_NAME_INT               int32  
 6   AIRLINE_AIRPORT_FLIGHTS_MONTH  int64  
 7   AVG_MONTHLY_PASS_AIRPORT       int64  
 8   AVG_MONTHLY_PASS_AIRLINE       int64  
 9   FLT_ATTENDANTS_PER_PASS        float64
 10  GROUND_SERV_PER_PASS           float64
 11  PLANE_AGE                      int64  
 12  LATITUDE                       float64
 13  LONGITUDE                      float64
 14  PRCP                           float64
 15  SNOW                           float64
 16  SNWD                           float64
 17  TMAX                           float64
 

In [ ]:
# df_validate.to_pickle("drive/MyDrive/Validation_Data_Flight_Delays.pkl")

In [ ]:
# df_test.to_pickle("drive/MyDrive/Train_Test_Data_Flight_Delays.pkl")

In [ ]:
df_unique_airports = pd.DataFrame(df[['LATITUDE','LONGITUDE']].value_counts()).reset_index()


In [ ]:
df_delay_imbalance = pd.DataFrame(columns = ['LAT','LNG','DELAYED','NOT-DELAYED'])
print(df_delay_imbalance)
for index,row in df_unique_airports[['LATITUDE','LONGITUDE']].iterrows():
  delayed = df.loc[(df['LATITUDE']==row['LATITUDE']) & (df['LONGITUDE']==row['LONGITUDE'])]['DEP_DEL15'].value_counts().iloc[1]
  not_delayed = df.loc[(df['LATITUDE']==row['LATITUDE']) & (df['LONGITUDE']==row['LONGITUDE'])]['DEP_DEL15'].value_counts().iloc[0]
  # df_delay_imbalance.append({'LAT': row['LATITUDE'],'LNG':row['LONGITUDE'],'DELAYED':delayed, 'NOT-DELAYED':not_delayed},ignore_index=True)
  # pd.DataFrame(columns = ['LAT','LNG','DELAYED','NOT-DELAYED'])
  df_delay_imbalance = pd.concat([df_delay_imbalance,pd.DataFrame([{'LAT':row['LATITUDE'],'LNG':row['LONGITUDE'],'DELAYED':delayed,'NOT-DELAYED':not_delayed}])],ignore_index=True)

print(df_delay_imbalance)
df['DEP_DEL15'].value_counts()

Empty DataFrame
Columns: [LAT, LNG, DELAYED, NOT-DELAYED]
Index: []
       LAT      LNG DELAYED NOT-DELAYED
0   33.641  -84.427   65892      326711
1   41.978  -87.906   74049      254996
2   32.894  -97.030   65497      230952
3   39.774 -104.880   55609      191566
4   35.219  -80.936   44958      186897
..     ...      ...     ...         ...
91  43.644  -70.304     790        2819
92  30.473  -87.188     745        2833
93  47.620 -117.533     455        2966
94  28.775  -81.240     341        1952
95  18.440  -66.002      23          61

[96 rows x 4 columns]


0    5261694
1    1227368
Name: DEP_DEL15, dtype: int64

In [ ]:
df_delay_imbalance['RATIO'] = df_delay_imbalance['DELAYED']/df_delay_imbalance['NOT-DELAYED']
print(df_delay_imbalance.sort_values(by='RATIO'))
lat = df_delay_imbalance.iloc[26]['LAT']
lng = df_delay_imbalance.loc[26]['LNG']

       LAT      LNG DELAYED NOT-DELAYED     RATIO
79  19.739 -156.046    1124       13002  0.086448
86  21.979 -159.346     939        9941  0.094457
69  61.169 -149.985    1870       16958  0.110272
38  21.319 -157.922    5265       46753  0.112613
53  20.901 -156.434    2937       24277  0.120979
..     ...      ...     ...         ...       ...
28  32.847  -96.853   16952       54671  0.310073
35  29.645  -95.279   14445       44079  0.327707
17  40.696  -74.172   33746       98732  0.341794
26  41.785  -87.752   21363       59140  0.361228
95  18.440  -66.002      23          61  0.377049

[96 rows x 5 columns]


In [ ]:
print(lat)
print(lng)

41.785
-87.752


## We pick the airport Chicago Midway International Airport at Latitude Longitude (41.785, -87.752)

In [ ]:
df_unique_airports.rename(columns = {0:'Count'},inplace=True)

In [ ]:
df_unique_airports = df_unique_airports.loc[df_unique_airports['Count']<100000]

In [ ]:
df_unique_airports.head()

LATITUDE  LONGITUDE  Count
22    26.074    -80.152  97175
23    32.733   -117.187  91942
24    25.792    -80.286  88185
25    36.127    -86.682  81289
26    41.785    -87.752  80503

In [ ]:
df_training_data = df.loc[(df['LATITUDE']==lat) & (df['LONGITUDE']==lng)]

In [ ]:
df_training_data['DEP_DEL15'].value_counts()

0    59140
1    21363
Name: DEP_DEL15, dtype: int64

### Creating the data splits

In [ ]:
y = df_training_data['DEP_DEL15']
X = df_training_data.drop('DEP_DEL15',axis = 1)
from sklearn.model_selection import train_test_split

X_train, X_validate, y_train, y_validate = train_test_split(X, y, test_size=0.30, random_state=2022, stratify = y)

df_validate = pd.concat([X_validate, y_validate], axis = 1)
df_test = pd.concat([X_train, y_train],axis = 1)


In [ ]:
df_validate.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24151 entries, 5869911 to 4821660
Data columns (total 20 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   MONTH                          24151 non-null  int64  
 1   DAY_OF_WEEK                    24151 non-null  int64  
 2   DISTANCE_GROUP                 24151 non-null  int64  
 3   SEGMENT_NUMBER                 24151 non-null  int64  
 4   CONCURRENT_FLIGHTS             24151 non-null  int64  
 5   CARRIER_NAME_INT               24151 non-null  int32  
 6   AIRLINE_AIRPORT_FLIGHTS_MONTH  24151 non-null  int64  
 7   AVG_MONTHLY_PASS_AIRPORT       24151 non-null  int64  
 8   AVG_MONTHLY_PASS_AIRLINE       24151 non-null  int64  
 9   FLT_ATTENDANTS_PER_PASS        24151 non-null  float64
 10  GROUND_SERV_PER_PASS           24151 non-null  float64
 11  PLANE_AGE                      24151 non-null  int64  
 12  LATITUDE                       24151 n

In [ ]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 56352 entries, 843141 to 5298251
Data columns (total 20 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   MONTH                          56352 non-null  int64  
 1   DAY_OF_WEEK                    56352 non-null  int64  
 2   DISTANCE_GROUP                 56352 non-null  int64  
 3   SEGMENT_NUMBER                 56352 non-null  int64  
 4   CONCURRENT_FLIGHTS             56352 non-null  int64  
 5   CARRIER_NAME_INT               56352 non-null  int32  
 6   AIRLINE_AIRPORT_FLIGHTS_MONTH  56352 non-null  int64  
 7   AVG_MONTHLY_PASS_AIRPORT       56352 non-null  int64  
 8   AVG_MONTHLY_PASS_AIRLINE       56352 non-null  int64  
 9   FLT_ATTENDANTS_PER_PASS        56352 non-null  float64
 10  GROUND_SERV_PER_PASS           56352 non-null  float64
 11  PLANE_AGE                      56352 non-null  int64  
 12  LATITUDE                       56352 no

In [ ]:
df_validate.reset_index(inplace=True)
df_test.reset_index(inplace=True)

In [ ]:
df_validate

index  MONTH  DAY_OF_WEEK  DISTANCE_GROUP  SEGMENT_NUMBER  \
0      5869911     11            4               1               3   
1      3137369      6            7               2               5   
2      5282839     10            2               2               4   
3       976252      3            1               2               5   
4       492502      1            2               2               2   
...        ...    ...          ...             ...             ...   
24146  4309420      8            7               1               2   
24147  5901985     11            7               7               4   
24148  5934188     11            7               4               4   
24149  2032582      4            1               4               1   
24150  4821660      9            6               7               2   

       CONCURRENT_FLIGHTS  CARRIER_NAME_INT  AIRLINE_AIRPORT_FLIGHTS_MONTH  \
0                      12                13                            248   
1                      16                14                           6607   
2                      13                14                           6890   
3                      19                14                           6398   
4                      10                14                           5259   
...                   ...               ...                            ...   
24146                  18                14                           6839   
24147                  17                14                           6044   
24148                  17                14                           6044   
24149                  11                14                           6362   
24150                  12                14                           6386   

       AVG_MONTHLY_PASS_AIRPORT  AVG_MONTHLY_PASS_AIRLINE  ...  \
0                        819756                   3472966  ...   
1                        819756                  13382999  ...   
2                        819756                  13382999  ...   
3                        819756                  13382999  ...   
4                        819756                  13382999  ...   
...                         ...                       ...  ...   
24146                    819756                  13382999  ...   
24147                    819756                  13382999  ...   
24148                    819756                  13382999  ...   
24149                    819756                  13382999  ...   
24150                    819756                  13382999  ...   

       GROUND_SERV_PER_PASS  PLANE_AGE  LATITUDE  LONGITUDE  PRCP  SNOW  SNWD  \
0                  0.000099          8    41.785    -87.752  0.00   0.1   0.0   
1                  0.000099         14    41.785    -87.752  0.20   0.0   0.0   
2                  0.000099         17    41.785    -87.752  0.00   0.0   0.0   
3                  0.000099         13    41.785    -87.752  0.00   0.0   0.0   
4                  0.000099          5    41.785    -87.752  0.18   0.5   5.1   
...                     ...        ...       ...        ...   ...   ...   ...   
24146              0.000099         19    41.785    -87.752  0.00   0.0   0.0   
24147              0.000099         12    41.785    -87.752  0.04   0.0   0.0   
24148              0.000099         18    41.785    -87.752  0.00   0.0   0.0   
24149              0.000099          7    41.785    -87.752  0.00   0.0   3.1   
24150              0.000099         14    41.785    -87.752  0.06   0.0   0.0   

       TMAX   AWND  DEP_DEL15  
0      31.0  11.18          0  
1      91.0   5.82          1  
2      86.0  10.07          0  
3      42.0  17.45          0  
4      33.0   9.40          0  
...     ...    ...        ...  
24146  82.0   7.38          0  
24147  40.0   6.49          0  
24148  50.0   9.40          0  
24149  52.0   8.28          0  
24150  80.0   7.61          0  

[24151 rows x 21 columns]

In [ ]:
df_test

index  MONTH  DAY_OF_WEEK  DISTANCE_GROUP  SEGMENT_NUMBER  \
0       843141      2            7               2               4   
1      5398935     10            6               4               2   
2      4661064      9            3               4               1   
3      2018389      4            6               7               2   
4       916026      2            7               3               3   
...        ...    ...          ...             ...             ...   
56347  2018435      4            6               3               4   
56348  4835097      9            4               2               1   
56349   822265      2            2               2               2   
56350  4617665      9            7               3               2   
56351  5298251     10            5               2               2   

       CONCURRENT_FLIGHTS  CARRIER_NAME_INT  AIRLINE_AIRPORT_FLIGHTS_MONTH  \
0                      15                14                           4769   
1                       8                14                           6890   
2                      13                14                           6386   
3                      16                14                           6362   
4                      12                14                           4769   
...                   ...               ...                            ...   
56347                  14                14                           6362   
56348                  16                14                           6386   
56349                  18                14                           4769   
56350                  11                14                           6386   
56351                  19                14                           6890   

       AVG_MONTHLY_PASS_AIRPORT  AVG_MONTHLY_PASS_AIRLINE  ...  \
0                        819756                  13382999  ...   
1                        819756                  13382999  ...   
2                        819756                  13382999  ...   
3                        819756                  13382999  ...   
4                        819756                  13382999  ...   
...                         ...                       ...  ...   
56347                    819756                  13382999  ...   
56348                    819756                  13382999  ...   
56349                    819756                  13382999  ...   
56350                    819756                  13382999  ...   
56351                    819756                  13382999  ...   

       GROUND_SERV_PER_PASS  PLANE_AGE  LATITUDE  LONGITUDE  PRCP  SNOW  SNWD  \
0                  0.000099         15    41.785    -87.752  0.02   0.0   0.0   
1                  0.000099         19    41.785    -87.752  2.18   0.0   0.0   
2                  0.000099          4    41.785    -87.752  0.00   0.0   0.0   
3                  0.000099          2    41.785    -87.752  0.00   0.0   0.0   
4                  0.000099         12    41.785    -87.752  0.20   3.1   0.0   
...                     ...        ...       ...        ...   ...   ...   ...   
56347              0.000099         15    41.785    -87.752  0.00   0.0   0.0   
56348              0.000099          5    41.785    -87.752  0.00   0.0   0.0   
56349              0.000099          9    41.785    -87.752  0.00   0.0   3.1   
56350              0.000099         15    41.785    -87.752  0.34   0.0   0.0   
56351              0.000099         22    41.785    -87.752  0.00   0.0   0.0   

       TMAX   AWND  DEP_DEL15  
0      50.0  24.38          1  
1      48.0  12.53          0  
2      69.0   8.95          0  
3      66.0   4.92          1  
4      27.0  11.18          1  
...     ...    ...        ...  
56347  66.0   4.92          0  
56348  74.0   4.92          0  
56349  35.0   4.70          0  
56350  73.0   4.47          0  
56351  59.0   6.93          0  

[56352 rows x 21 columns]

### Writing to files

In [ ]:
df_validate.to_pickle("drive/MyDrive/Validation_Data_Airport_Flight_Delays.pkl")

In [ ]:
df_test.to_pickle("drive/MyDrive/Train_Test_Data_Airport_Flight_Delays.pkl")